src/Simulation/Simulators/CommonNativeSimulator/CommonNativeSimulator.cs

```diff
@@ -0,0 +1,191 @@
+            }
+
+            used[q.Id] = true;
+        }
+
+        /// <summary>
+        ///     Makes sure the angle for a rotation or exp operation is not NaN or Infinity.
+        /// </summary>
+        static void CheckAngle(double angle)
+        {
+            IgnorableAssert.Assert(!(double.IsNaN(angle) || double.IsInfinity(angle)), "Invalid angle for rotation/exponentiation.");
+
+            if (double.IsNaN(angle)) throw new ArgumentOutOfRangeException("angle", "Angle can't be NaN.");
+            if (double.IsInfinity(angle)) throw new ArgumentOutOfRangeException("angle", "Angle can't be Infity.");
+        }
+
+        /// <summary>
+        ///     Makes sure the target qubit of an operation is valid. In particular it checks that the qubit instance is not null.
+        ///     Also sets the isMeasured flag to false for each qubit
+        /// </summary>
+        void CheckQubit(Qubit q1)
+        {
+            if (q1 == null) throw new ArgumentNullException(nameof(q1), "Trying to perform a primitive operation on a null Qubit");
+            //setting qubit as not measured to not allow release in case of gate operation on qubit
+            q1.IsMeasured = false;
+        }
+
+        /// <summary>
+        ///     Makes sure all qubits are valid as parameter of an intrinsic quantum operation. In particular it checks that 
+        ///         - none of the qubits are null
+        ///         - there are no duplicated qubits
+        ///     Also sets the isMeasured flag to false for each qubit
+        /// </summary>
+        bool[] CheckQubits(IQArray<Qubit> ctrls, Qubit q1)
+        {
+            bool[] used = new bool[((QSimQubitManager)QubitManager).MaxId];
+
+            CheckQubitInUse(q1, used);
+            q1.IsMeasured = false;
+
+            if (ctrls != null && ctrls.Length > 0)
+            {
+                foreach (var q in ctrls)
+                {
+                    CheckQubitInUse(q, used);
+                    //setting qubit as not measured to not allow release in case of gate operation on qubit
+                    q.IsMeasured = false;
+                }
+            }
+
+            return used;
+        }
+
+
+        /// <summary>
+        ///     Makes sure all qubits are valid as parameter of an intrinsic quantum operation. In particular it checks that 
+        ///         - none of the qubits are null
+        ///         - there are no duplicated qubits
+        ///     Also sets the isMeasured flag to false for each qubit
+        /// </summary>
+        bool[] CheckQubits(IQArray<Qubit> targets)
+        {
+            if (targets == null) throw new ArgumentNullException(nameof(targets), "Trying to perform an intrinsic operation on a null Qubit array.");
+            if (targets.Length == 0) throw new ArgumentNullException(nameof(targets), "Trying to perform an intrinsic operation on an empty Qubit array.");
+
+            bool[] used = new bool[((QSimQubitManager)QubitManager).MaxId];
```

Comment:
Here and in a few other places. For the full state simulator the maximum number of qubits were small. Sparse simulator can handle quite a few more in certain circumstances. It may not be always desirable to allocate array with the size of the number of qubits. I would consider array for a full state simulator and use HashSet for the sparse simulator.

Reply:
I didn't touch the SparseSimulator for a few months and I don't remember exactly its difference from the QuantumSimulator. And before I get back to the SparseSimulator I also need to resolve the Dump functionality issue. So paying significant attention to the SparseSim at this stage seems to me less efficient than if I move forward now. I will get back to SparseSim later any way, and that's when I plan to finalize the changes.
If there are no critical objections, I would prefer to move forward with this PR at the moment.

Reply:
I'm OK leaving this as is, especially given that sparse simulator won't be used on more than a few thousand qubits. Maybe we should note this somewhere - comments or work item.